internal/app/machined/pkg/runtime/v1alpha1/platform/metal/metal.go

```diff
@@ -94,24 +151,101 @@ func PopulateURLParameters(downloadURL string, getSystemUUID func() (string, err
 	return u.String(), nil
 }
 
-func getSystemUUID() (string, error) {
-	s, err := smbios.New()
+func getResource[T resource.Resource](ctx context.Context, r state.State, namespace, typ string, checkAndGetFunc func(T) string) (string, error) {
+	metadata := resource.NewMetadata(namespace, typ, "", resource.VersionUndefined)
+
+	watchCtx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	events := make(chan safe.WrappedStateEvent[T])
+
+	err := safe.StateWatchKind[T](watchCtx, r, metadata, events, state.WithBootstrapContents(true))
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to watch %s resources: %w", typ, err)
 	}
 
-	return s.SystemInformation.UUID, nil
+	var watchErr error
+
+	for {
+		select {
+		case <-watchCtx.Done():
+			err := fmt.Errorf("failed to determine %s: %w", typ, watchCtx.Err())
+			err = fmt.Errorf("%s; %w", err.Error(), watchErr)
+
+			return "", err
+		case event := <-events:
+			eventResource, err := event.Resource()
+			if err != nil {
+				watchErr = fmt.Errorf("%s; invalid resource in wrapped event: %w", watchErr.Error(), err)
+			}
+
+			val := checkAndGetFunc(eventResource)
+			if val != "" {
+				return val, nil
+			}
+		}
+	}
+}
+
+func getAndCheckUUID(r *hardwareResource.SystemInformation) string {
+	sysInfo := r.TypedSpec()
+	if sysInfo != nil {
```

Comment:
I don't think Spec can ever be `nil`, pointer is returned to allow modifications, but resource hold spec by value